pkg/json/snapshot.go

```diff
@@ -515,6 +518,203 @@ func translate(data any) (*snapshotReader, int64, error) {
+			name = upsertKeys[upsertIdx]
+			upsertOffsets = append(upsertOffsets, i)
+			upsertIdx++
+		default:
+			// Otherwise, compare current keys of original and upsert sets.
+			originalKey := originalKeys[origIdx]
+			upsertKey := upsertKeys[upsertIdx]
+			switch {
+			case upsertKey < originalKey:
+				name = upsertKey
+				upsertOffsets = append(upsertOffsets, i)
+				upsertIdx++
+			case upsertKey > originalKey:
+				name = originalKey
+				origIdx++
+			default: // equals case (upsert key wins)
+				name = upsertKey
+				upsertOffsets = append(upsertOffsets, i)
+				upsertIdx++
+				origIdx++
+			}
+		}
+		writeString(name, buffer)
+		order.PutUint32(buffer.Bytes()[keyOffsets:keyOffsets+4], uint32(nameOffset))
+		keyOffsets += 4
+	}
+
+	// ------------------------------------------------------------------------------------
+	// Incremental walk through properties of the object and upserts
+	// ------------------------------------------------------------------------------------
+	// Write the property values, updating the offset placeholders after each.
+	seenUpsertsIdx := 0
+	upsertIdx = 0
+	origIdx = 0
+	for i := range finalLength {
+		var element any
+
+		// Done with all upserts?
+		if seenUpsertsIdx == len(upsertOffsets) {
+			// Done with upserts.
+			element = original.iterate(origIdx)
+			origIdx++
+		} else if i == upsertOffsets[seenUpsertsIdx] {
+			// Known upsert? Slap in the value.
+			element = upsertEntries[upsertKeys[upsertIdx]]
+			if upsertOffsets[seenUpsertsIdx] < len(originalKeys) {
+				origIdx++ // Skip this entry from the originals, it was replaced.
+			}
+			upsertIdx++
+			seenUpsertsIdx++
+		} else {
+			// Not an upsert.
+			element = original.iterate(origIdx)
+			origIdx++
+		}
+
+		valueOffset, err := serialize(element, cache, buffer, base)
+		if err != nil {
+			return offset, err
+		}
+
+		order.PutUint32(buffer.Bytes()[valueOffsets:valueOffsets+4], uint32(valueOffset))
+		valueOffsets += 4
+	}
+	// ------------------------------------------------------------------------------------
+	return offset, nil
+}
+
+// UpsertObjectPatches writes out an ObjectBinary containing the original object,
+// plus the equivalent of JSON Patch "add" operations (upserts), using the
+// provided key/value slices. It's intended to emulate a subset of the OPA
+// EditTree data structure's functionality.
+func UpsertObjectPatches(original Object, upsertEntries map[string]any, buffer *bytes.Buffer, base int32) (int32, error) {
+	cache := newEncodingCache()
+	newItemCount := 0
+	preexistingItemCount := 0
+	for k := range upsertEntries {
+		if _, err := original.Extract("/" + k); err != nil {
+			newItemCount++
+		} else {
+			preexistingItemCount++
+		}
+	}
+
+	// Sorted list of upsert keys, useful during properties list building.
+	upsertKeys := make([]string, 0, len(upsertEntries))
+	for k := range upsertEntries {
+		upsertKeys = append(upsertKeys, k)
+	}
+	slices.Sort(upsertKeys)
```

Comment:
💭 I was wondering if there was an alternative way here using `maps.Keys()` and some other type of sort method, but I don't think it'll give us much over what we have.